src/Csv.elm

```diff
@@ -83,100 +118,140 @@ thrd ( _, _, c ) =
 doubleDoubleQuote =
-    (doubleQuote *> doubleQuote)
+    doubleQuote |. doubleQuote
+
+
+textData : Char -> Parser ()
+textData sepChar =
+    chompIf <| textChar sepChar
+
+
+textChar : Char -> Char -> Bool
+textChar sepChar c =
+    not (List.member c [ '"', sepChar, '\n', crc ])
 
 
+nonEscaped : Char -> Parser String
+nonEscaped sepChar =
+    getChompedString (chompWhile (textChar sepChar))
 
--- Grab all non-quote data.
 
+innerChar : Char -> Parser String
+innerChar sepChar =
+    Parser.map (String.replace "\"\"" "\"") <|
+        backtrackable <|
+            getChompedString
+                (oneOf [ textData sepChar, comma, cr, lf, doubleDoubleQuote ])
 
-textData : Config -> Parser s Char
-textData config =
-    noneOf [ '"', config.fieldSep , '\n', '\x0D' ]
 
+innerString : Char -> List String -> Parser (Step (List String) String)
+innerString sepChar strs =
+    oneOf
+        [ succeed (\str -> Loop (str :: strs)) |= innerChar sepChar
+        , succeed ()
+            |> Parser.map (\_ -> Done (String.concat (List.reverse strs)))
+        ]
 
-nonEscaped : Config -> Parser s String
-nonEscaped config =
-    Combine.many (textData config)
-        |> Combine.map String.fromList
-        |> Combine.mapError (always [ "Expected non-escaped value." ])
 
+escaped : Char -> Parser String
+escaped sepChar =
+    succeed identity
+        |. doubleQuote
+        |= loop [] (innerString sepChar)
+        |. doubleQuote
 
-escaped : Config -> Parser s String
-escaped config =
-    let
-        innerChar =
-            Combine.choice [ textData config, char config.fieldSep, cr, lf, doubleDoubleQuote ]
 
-        innerString =
-            Combine.many innerChar |> Combine.map String.fromList
-    in
-        (doubleQuote *> innerString <* doubleQuote)
-            |> Combine.mapError (always [ "Expected escaped value." ])
+field : Char -> Parser String
+field sepChar =
+    oneOf [ escaped sepChar, nonEscaped sepChar ]
 
 
-field : Config -> Parser s String
-field config =
-    Combine.choice [ escaped config, nonEscaped config ]
+name : Char -> Parser String
+name sepChar =
+    field sepChar
 
 
-name : Config -> Parser s String
-name =
-    field
+recordHelper : Char -> List String -> Parser (Step (List String) (List String))
+recordHelper sepChar strs =
+    oneOf
+        [ backtrackable <|
+            succeed (\str -> Loop (str :: strs))
+                |= field sepChar
+                |. symbol (String.fromChar sepChar)
+        , succeed (\str -> Done (List.reverse (str :: strs)))
+            |= field sepChar
+            |. lineSep
+        ]
 
 
-header : Config -> Parser s (List String)
-header = record
+record : Char -> Parser (List String)
+record sepChar =
+    loop [] (recordHelper sepChar)
 
 
-record : Config -> Parser s (List String)
-record config =
-    Combine.sepBy (char config.fieldSep) (field config)
+recordsHelper : Char -> List (List String) -> Parser (Step (List (List String)) (List (List String)))
```

Comment:
With both `recordHelper` and `recordsHelper` I can't help but wonder if we could choose some better names.  Is it possible we just want to make these local functions of the associated functions in a `let` binding?  That would avoid having to have a module-wide name for such a local function.

Also, the type signature is so long that it really makes me wonder if it could be simplified.